File: AssetSync/src/AssetSync.Core/Services/SyncEngine.cs
using AssetSync.Core.Enums;
using AssetSync.Core.Interfaces;
using AssetSync.Core.Models;
using Microsoft.Extensions.Logging;
using LogLevel = AssetSync.Core.Enums.LogLevel;

namespace AssetSync.Core.Services;

public class SyncEngine : ISyncEngine
{
    private readonly IIntuneService _intuneService;
    private readonly IIruService _iruService;
    private readonly ISnipeItService _snipeItService;
    private readonly ILogRepository _logRepository;
    private readonly IMappingRepository _mappingRepository;
    private readonly IConfigRepository _configRepository;
    private readonly IWebhookService _webhookService;
    private readonly IConnectivityTester _connectivityTester;
    private readonly DeviceMerger _merger;
    private readonly ConflictResolver _resolver;
    private readonly BuildVersionMapper _buildMapper;
    private readonly ILogger<SyncEngine> _logger;

    public SyncEngine(
        IIntuneService intuneService,
        IIruService iruService,
        ISnipeItService snipeItService,
        ILogRepository logRepository,
        IMappingRepository mappingRepository,
        IConfigRepository configRepository,
        IWebhookService webhookService,
        IConnectivityTester connectivityTester,
        DeviceMerger merger,
        ConflictResolver resolver,
        BuildVersionMapper buildMapper,
        ILogger<SyncEngine> logger)
    {
        _intuneService = intuneService;
        _iruService = iruService;
        _snipeItService = snipeItService;
        _logRepository = logRepository;
        _mappingRepository = mappingRepository;
        _configRepository = configRepository;
        _webhookService = webhookService;
        _connectivityTester = connectivityTester;
        _merger = merger;
        _resolver = resolver;
        _buildMapper = buildMapper;
        _logger = logger;
    }

    public async Task<SyncRunSummary> RunSyncAsync(bool dryRun, CancellationToken cancellationToken = default)
    {
        var runId = Guid.NewGuid().ToString();
        var summary = new SyncRunSummary
        {
            SyncRunId = runId,
            StartedAtUtc = DateTimeOffset.UtcNow,
            DryRun = dryRun
        };

        await LogAsync(runId, LogLevel.Info, SourceSystem.Application, dryRun ? "sync_start_dry" : "sync_start", null, null, true, null, cancellationToken).ConfigureAwait(false);

        try
        {
            var snipeStatus = await _connectivityTester.TestSnipeItAsync(cancellationToken).ConfigureAwait(false);
            summary.SnipeItReachable = snipeStatus.State == ConnectionState.Connected;
            if (!summary.SnipeItReachable)
            {
                await _webhookService.SendConnectivityFailureNotificationAsync("Snipe-IT", snipeStatus.Message ?? "Unreachable", cancellationToken).ConfigureAwait(false);
                summary.CompletedAtUtc = DateTimeOffset.UtcNow;
                await LogAsync(runId, LogLevel.Error, SourceSystem.Application, "sync_complete", null, null, false, "Snipe-IT unreachable", cancellationToken).ConfigureAwait(false);
                await _webhookService.SendSyncNotificationAsync(summary, cancellationToken).ConfigureAwait(false);
                return summary;
            }

            var intuneStatus = await _connectivityTester.TestIntuneAsync(cancellationToken).ConfigureAwait(false);
            summary.IntuneReachable = intuneStatus.State == ConnectionState.Connected;
            var iruStatus = await _connectivityTester.TestIruAsync(cancellationToken).ConfigureAwait(false);
            summary.IruReachable = iruStatus.State == ConnectionState.Connected;

            if (!summary.IntuneReachable && !summary.IruReachable)
            {
                await _webhookService.SendConnectivityFailureNotificationAsync("Intune and Iru", "Both MDM sources unreachable", cancellationToken).ConfigureAwait(false);
                summary.CompletedAtUtc = DateTimeOffset.UtcNow;
                await LogAsync(runId, LogLevel.Error, SourceSystem.Application, "sync_complete", null, null, false, "No MDM source available", cancellationToken).ConfigureAwait(false);
                await _webhookService.SendSyncNotificationAsync(summary, cancellationToken).ConfigureAwait(false);
                return summary;
            }

            var intuneList = summary.IntuneReachable
                ? await _intuneService.GetManagedDevicesAsync(cancellationToken).ConfigureAwait(false)
                : Array.Empty<Device>();
            var iruList = summary.IruReachable
                ? await _iruService.GetDevicesAsync(cancellationToken).ConfigureAwait(false)
                : Array.Empty<Device>();

            var merged = _merger.Merge(intuneList, iruList);

            foreach (var device in merged)
            {
                if (string.IsNullOrEmpty(device.NormalizedSerial))
                    device.NormalizedSerial = SerialNumberNormalizer.Normalize(device.SerialNumber);

                var existing = await _snipeItService.SearchAssetsBySerialAsync(device.NormalizedSerial, cancellationToken).ConfigureAwait(false);
                if (existing.Count > 1)
                {
                    summary.Errors++;
                    await LogAsync(runId, LogLevel.Error, SourceSystem.SnipeIt, "error", device.SerialNumber, device.DeviceName, false, "Multiple Snipe-IT assets with same serial", cancellationToken).ConfigureAwait(false);
                    continue;
                }

                if (existing.Count == 0)
                {
                    var modelMapping = await _mappingRepository.GetModelMappingAsync(device.Model ?? "", cancellationToken).ConfigureAwait(false);
                    if (modelMapping == null)
                    {
                        summary.Skipped++;
                        await LogAsync(runId, LogLevel.Warning, SourceSystem.Application, "skip", device.SerialNumber, device.DeviceName, true, "Pending model mapping", cancellationToken).ConfigureAwait(false);
                        continue;
                    }
                    device.SnipeItModelId = modelMapping.SnipeItModelId;
                    device.WindowsFeatureUpdate = await _buildMapper.GetFriendlyNameAsync(device.OsVersion, cancellationToken).ConfigureAwait(false);
                    var catMapping = await _mappingRepository.GetCategoryMappingAsync(device.DeviceType ?? "", cancellationToken).ConfigureAwait(false);
                    if (catMapping != null) device.SnipeItCategoryId = catMapping.SnipeItCategoryId;
                    if (!dryRun)
                    {
                        var created = await _snipeItService.CreateAssetAsync(device, cancellationToken).ConfigureAwait(false);
                        if (created != null)
                        {
                            summary.Created++;
                            await LogAsync(runId, LogLevel.Info, SourceSystem.SnipeIt, "create", device.SerialNumber, device.DeviceName, true, null, cancellationToken).ConfigureAwait(false);
                        }
                        else
                        {
                            summary.Errors++;
                            await LogAsync(runId, LogLevel.Error, SourceSystem.SnipeIt, "error", device.SerialNumber, device.DeviceName, false, "Create failed", cancellationToken).ConfigureAwait(false);
                        }
                    }
                    else
                    {
                        summary.Created++;
                        await LogAsync(runId, LogLevel.Info, SourceSystem.Application, "create", device.SerialNumber, device.DeviceName, true, "[DRY RUN]", cancellationToken).ConfigureAwait(false);
                    }
                    continue;
                }

                var snipeAsset = existing[0];
                var updates = _resolver.GetUpdatesToApply(snipeAsset, device);
                var discrepancies = _resolver.GetDiscrepancies(snipeAsset, device);
                foreach (var (field, snipeVal, mdmVal) in discrepancies)
                    await LogAsync(runId, LogLevel.Warning, SourceSystem.Application, "skip", device.SerialNumber, device.DeviceName, true, $"Discrepancy {field}: Snipe-IT={snipeVal} MDM={mdmVal}", cancellationToken).ConfigureAwait(false);

                if (updates.Count == 0)
                {
                    summary.Skipped++;
                    continue;
                }

                if (!dryRun && snipeAsset.SnipeItAssetId != null)
                {
                    var ok = await _snipeItService.UpdateAssetAsync(snipeAsset.SnipeItAssetId.Value, updates, cancellationToken).ConfigureAwait(false);
                    if (ok)
                    {
                        summary.Updated++;
                        await LogAsync(runId, LogLevel.Info, SourceSystem.SnipeIt, "update", device.SerialNumber, device.DeviceName, true, null, cancellationToken).ConfigureAwait(false);
                    }
                    else
                    {
                        summary.Errors++;
                        await LogAsync(runId, LogLevel.Error, SourceSystem.SnipeIt, "error", device.SerialNumber, device.DeviceName, false, "Update failed", cancellationToken).ConfigureAwait(false);
                    }
                }
                else
                {
                    summary.Updated++;
                    await LogAsync(runId, LogLevel.Info, SourceSystem.Application, "update", device.SerialNumber, device.DeviceName, true, dryRun ? "[DRY RUN]" : null, cancellationToken).ConfigureAwait(false);
                }
            }

            summary.CompletedAtUtc = DateTimeOffset.UtcNow;
            await LogAsync(runId, LogLevel.Info, SourceSystem.Application, "sync_complete", null, null, true, null, cancellationToken).ConfigureAwait(false);
            await _webhookService.SendSyncNotificationAsync(summary, cancellationToken).ConfigureAwait(false);

            var retention = TimeSpan.FromDays(30);
            await _logRepository.PurgeOlderThanAsync(retention, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Sync failed");
            summary.CompletedAtUtc = DateTimeOffset.UtcNow;
            summary.Errors++;
            await LogAsync(runId, LogLevel.Error, SourceSystem.Application, "sync_complete", null, null, false, ex.Message, cancellationToken).ConfigureAwait(false);
            await _webhookService.SendSyncNotificationAsync(summary, cancellationToken).ConfigureAwait(false);
        }

        return summary;
    }

    private async Task LogAsync(string runId, LogLevel level, SourceSystem source, string action, string? serial, string? deviceName, bool success, string? errorDetail, CancellationToken cancellationToken)
    {
        await _logRepository.AppendAsync(new LogEntry
        {
            TimestampUtc = DateTimeOffset.UtcNow,
            Level = level,
            SourceSystem = source,
            Action = action,
            SerialNumber = serial,
            DeviceName = deviceName,
            Success = success,
            ErrorDetail = errorDetail,
            SyncRunId = runId
        }, cancellationToken).ConfigureAwait(false);
    }
}

File: AssetSync-macOS/AssetSyncCore/Sources/AssetSyncCore/Services/SyncEngine.swift
import Foundation
import os

/// Main sync orchestrator — coordinates the full sync cycle across
/// Intune, Iru, and Snipe-IT.
public actor SyncEngine: SyncEngineProtocol {
    private let intuneService: IntuneServiceProtocol
    private let iruService: IruServiceProtocol
    private let snipeItService: SnipeItServiceProtocol
    private let logRepository: LogRepositoryProtocol
    private let mappingRepository: MappingRepositoryProtocol
    private let configRepository: ConfigRepositoryProtocol
    private let webhookService: WebhookServiceProtocol
    private let connectivityTester: ConnectivityTesterProtocol
    private let merger: DeviceMerger
    private let resolver: ConflictResolver
    private let buildMapper: BuildVersionMapper
    private let logger = Logger(subsystem: "com.assetsync", category: "SyncEngine")

    /// Matches a standard PM asset tag: PM followed by exactly 7 digits.
    private static let pmTagRegex = /^PM\d{7}$/

    private static func isStandardTag(_ tag: String?) -> Bool {
        guard let tag, !tag.isEmpty else { return false }
        return tag.wholeMatch(of: pmTagRegex.ignoresCase()) != nil
    }

    public init(
        intuneService: IntuneServiceProtocol,
        iruService: IruServiceProtocol,
        snipeItService: SnipeItServiceProtocol,
        logRepository: LogRepositoryProtocol,
        mappingRepository: MappingRepositoryProtocol,
        configRepository: ConfigRepositoryProtocol,
        webhookService: WebhookServiceProtocol,
        connectivityTester: ConnectivityTesterProtocol,
        merger: DeviceMerger,
        resolver: ConflictResolver,
        buildMapper: BuildVersionMapper
    ) {
        self.intuneService = intuneService
        self.iruService = iruService
        self.snipeItService = snipeItService
        self.logRepository = logRepository
        self.mappingRepository = mappingRepository
        self.configRepository = configRepository
        self.webhookService = webhookService
        self.connectivityTester = connectivityTester
        self.merger = merger
        self.resolver = resolver
        self.buildMapper = buildMapper
    }

    public func runSync(dryRun: Bool) async throws -> SyncRunSummary {
        let runId = UUID().uuidString
        var summary = SyncRunSummary()
        summary.syncRunId = runId
        summary.startedAtUtc = Date()
        summary.dryRun = dryRun

        await log(runId: runId, level: .info, source: .application,
                  action: dryRun ? "sync_start_dry" : "sync_start")

        do {
            // 1. Connectivity checks
            let snipeStatus = await connectivityTester.testSnipeIt()
            summary.snipeItReachable = snipeStatus.state == .connected
            guard summary.snipeItReachable else {
                await webhookService.sendConnectivityFailureNotification(
                    serviceName: "Snipe-IT", message: snipeStatus.message ?? "Unreachable")
                summary.completedAtUtc = Date()
                await log(runId: runId, level: .error, source: .application,
                          action: "sync_complete", errorDetail: "Snipe-IT unreachable")
                await webhookService.sendSyncNotification(summary)
                return summary
            }

            let intuneStatus = await connectivityTester.testIntune()
            summary.intuneReachable = intuneStatus.state == .connected
            let iruStatus = await connectivityTester.testIru()
            summary.iruReachable = iruStatus.state == .connected

            guard summary.intuneReachable || summary.iruReachable else {
                await webhookService.sendConnectivityFailureNotification(
                    serviceName: "Intune and Iru", message: "Both MDM sources unreachable")
                summary.completedAtUtc = Date()
                await log(runId: runId, level: .error, source: .application,
                          action: "sync_complete", errorDetail: "No MDM source available")
                await webhookService.sendSyncNotification(summary)
                return summary
            }

            // 2. Fetch devices
            let intuneList = summary.intuneReachable ? (try? await intuneService.getManagedDevices()) ?? [] : []
            let iruList = summary.iruReachable ? (try? await iruService.getDevices()) ?? [] : []

            let writeBackIntune = (try? await configRepository.getWriteBackIntuneEnabled()) ?? false
            let writeBackIru = (try? await configRepository.getWriteBackIruEnabled()) ?? false
            let intuneMdmWins = (try? await configRepository.getIntuneMdmWins()) ?? false
            let iruMdmWins = (try? await configRepository.getIruMdmWins()) ?? false

            let ignoredModels = Set((try? await mappingRepository.getIgnoredModels()) ?? [])

            // 3. Merge
            let merged = merger.merge(intuneDevices: intuneList, iruDevices: iruList)

            // 4. Process each device
            for var device in merged {
                if device.normalizedSerial.isEmpty {
                    device.normalizedSerial = SerialNumberNormalizer.normalize(device.serialNumber)
                }

                let existing = (try? await snipeItService.searchAssetsBySerial(device.normalizedSerial)) ?? []

                if existing.count > 1 {
                    summary.errors += 1
                    await log(runId: runId, level: .error, source: .snipeIt, action: "error",
                              serial: device.serialNumber, deviceName: device.deviceName,
                              success: false, errorDetail: "Multiple Snipe-IT assets with same serial")
                    continue
                }

                if existing.isEmpty {
                    // Create flow
                    guard let modelMapping = try? await mappingRepository.getModelMapping(
                        mdmModelString: device.model ?? "") else {
                        summary.skipped += 1
                        if !ignoredModels.contains(device.model?.lowercased() ?? "") {
                            await log(runId: runId, level: .warning, source: .application, action: "skip",
                                      serial: device.serialNumber, deviceName: device.deviceName,
                                      errorDetail: "Pending model mapping: \(device.model ?? "")")
                        }
                        continue
                    }
                    device.snipeItModelId = modelMapping.snipeItModelId
                    device.windowsFeatureUpdate = await buildMapper.getFriendlyName(for: device.osVersion)
                    if let catMapping = try? await mappingRepository.getCategoryMapping(
                        mdmDeviceType: device.deviceType ?? "") {
                        device.snipeItCategoryId = catMapping.snipeItCategoryId
                    }
                    if Self.isStandardTag(device.mdmAssetTag) {
                        device.mdmAssetTag = device.mdmAssetTag?.uppercased()
                    }
                    if !dryRun {
                        do {
                            if let created = try await snipeItService.createAsset(device) {
                                summary.created += 1
                                await log(runId: runId, level: .info, source: .snipeIt, action: "create",
                                          serial: device.serialNumber, deviceName: device.deviceName)
                                await writeBackAssetTag(runId: runId, device: device,
                                                        assetTag: created.snipeItAssetTag,
                                                        writeBackIntune: writeBackIntune,
                                                        writeBackIru: writeBackIru)
                            } else {
                                summary.errors += 1
                                await log(runId: runId, level: .error, source: .snipeIt, action: "error",
                                          serial: device.serialNumber, deviceName: device.deviceName,
                                          success: false, errorDetail: "Create failed")
                            }
                        } catch {
                            summary.errors += 1
                            await log(runId: runId, level: .error, source: .snipeIt, action: "error",
                                      serial: device.serialNumber, deviceName: device.deviceName,
                                      success: false, errorDetail: error.localizedDescription)
                        }
                    } else {
                        summary.created += 1
                        await log(runId: runId, level: .info, source: .application, action: "create",
                                  serial: device.serialNumber, deviceName: device.deviceName,
                                  errorDetail: "[DRY RUN]")
                    }
                    continue
                }

                // Update flow
                let snipeAsset = existing[0]
                let mdmWins: Bool
                if device.platformSource == "Intune" { mdmWins = intuneMdmWins }
                else if device.platformSource == "Iru" { mdmWins = iruMdmWins }
                else { mdmWins = device.iruDeviceId?.isEmpty == false ? iruMdmWins : intuneMdmWins }

                let resolvedUpdates = resolver.getUpdatesToApply(
                    snipeItAsset: snipeAsset, mdmDevice: device, mdmWins: mdmWins)
                let discrepancies = resolver.getDiscrepancies(
                    snipeItAsset: snipeAsset, mdmDevice: device)
                for d in discrepancies {
                    await log(runId: runId, level: .warning, source: .application, action: "skip",
                              serial: device.serialNumber, deviceName: device.deviceName,
                              errorDetail: "Discrepancy \(d.field): Snipe-IT=\(d.snipeItValue ?? "") MDM=\(d.mdmValue ?? "")")
                }

                var updates = resolvedUpdates
                if Self.isStandardTag(device.mdmAssetTag) && !Self.isStandardTag(snipeAsset.snipeItAssetTag) {
                    updates["asset_tag"] = device.mdmAssetTag!.uppercased()
                    await log(runId: runId, level: .info, source: .application, action: "asset_tag_push",
                              serial: device.serialNumber, deviceName: device.deviceName,
                              errorDetail: "MDM asset tag '\(device.mdmAssetTag ?? "")' → Snipe-IT (was: '\(snipeAsset.snipeItAssetTag ?? "none")')")
                }

                guard !updates.isEmpty else {
                    summary.skipped += 1
                    continue
                }

                if !dryRun, let assetId = snipeAsset.snipeItAssetId {
                    let ok = (try? await snipeItService.updateAsset(assetId: assetId, updates: updates)) ?? false
                    if ok {
                        summary.updated += 1
                        await log(runId: runId, level: .info, source: .snipeIt, action: "update",
                                  serial: device.serialNumber, deviceName: device.deviceName)
                        await writeBackAssetTag(runId: runId, device: device,
                                                assetTag: snipeAsset.snipeItAssetTag,
                                                writeBackIntune: writeBackIntune,
                                                writeBackIru: writeBackIru)
                    } else {
                        summary.errors += 1
                        await log(runId: runId, level: .error, source: .snipeIt, action: "error",
                                  serial: device.serialNumber, deviceName: device.deviceName,
                                  success: false, errorDetail: "Update failed")
                    }
                } else {
                    summary.updated += 1
                    await log(runId: runId, level: .info, source: .application, action: "update",
                              serial: device.serialNumber, deviceName: device.deviceName,
                              errorDetail: dryRun ? "[DRY RUN]" : nil)
                }
            }

            summary.completedAtUtc = Date()
            await log(runId: runId, level: .info, source: .application, action: "sync_complete")
            await webhookService.sendSyncNotification(summary)

            try? await logRepository.purgeOlderThan(30 * 24 * 60 * 60) // 30 days
        } catch {
            logger.error("Sync failed: \(error.localizedDescription)")
            summary.completedAtUtc = Date()
            summary.errors += 1
            await log(runId: runId, level: .error, source: .application, action: "sync_complete",
                      success: false, errorDetail: "\(error.localizedDescription)\n\(String(describing: error))")
            await webhookService.sendSyncNotification(summary)
        }

        return summary
    }

    // MARK: - Private helpers

    private func writeBackAssetTag(runId: String, device: Device, assetTag: String?,
                                    writeBackIntune: Bool, writeBackIru: Bool) async {
        guard let assetTag, !assetTag.isEmpty,
              assetTag.wholeMatch(of: Self.pmTagRegex.ignoresCase()) != nil else { return }

        if writeBackIntune, let intuneId = device.intuneDeviceId, !intuneId.isEmpty,
           device.platformSource == "Intune" {
            let alreadyInNotes = device.intuneNotes?.localizedCaseInsensitiveContains("Asset Tag: \(assetTag)") ?? false
            if !alreadyInNotes {
                do {
                    _ = try await intuneService.writeBackAssetTag(
                        intuneDeviceId: intuneId, assetTag: assetTag, existingNotes: device.intuneNotes)
                    await log(runId: runId, level: .info, source: .intune, action: "write_back",
                              serial: device.serialNumber, deviceName: device.deviceName,
                              errorDetail: "Asset tag '\(assetTag)' written to Intune notes")
                } catch {
                    await log(runId: runId, level: .warning, source: .intune, action: "write_back",
                              serial: device.serialNumber, deviceName: device.deviceName,
                              success: false, errorDetail: error.localizedDescription)
                }
            }
        }

        if writeBackIru, let iruId = device.iruDeviceId, !iruId.isEmpty,
           device.platformSource == "Iru" {
            guard !assetTag.caseInsensitiveCompare(device.mdmAssetTag ?? "").rawValue.signum() == 0 else { return }
            let ok = (try? await iruService.writeBackAssetTag(iruDeviceId: iruId, assetTag: assetTag)) ?? false
            await log(runId: runId, level: ok ? .info : .warning, source: .iru, action: "write_back",
                      serial: device.serialNumber, deviceName: device.deviceName,
                      success: ok, errorDetail: ok ? "Asset tag '\(assetTag)' written to Iru" : "Write-back to Iru failed")
        }
    }

    private func log(runId: String, level: LogLevel, source: SourceSystem, action: String,
                     serial: String? = nil, deviceName: String? = nil,
                     success: Bool = true, errorDetail: String? = nil) async {
        var entry = LogEntry()
        entry.timestampUtc = Date()
        entry.level = level
        entry.sourceSystem = source
        entry.action = action
        entry.serialNumber = serial
        entry.deviceName = deviceName
        entry.success = success
        entry.errorDetail = errorDetail
        entry.syncRunId = runId
        try? await logRepository.append(entry)
    }
}
